pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -173,32 +221,63 @@ func (m *mergeQuerier) mergeDistinctStringSlice(f stringSliceFunc) ([]string, st
 func (m *mergeQuerier) Close() error {
 	errs := tsdb_errors.NewMulti()
 	for pos, tenantID := range m.tenantIDs {
-		errs.Add(errors.Wrapf(m.queriers[pos].Close(), "failed to close querier for tenant id %s", tenantID))
+		errs.Add(errors.Wrapf(m.queriers[pos].Close(), `failed to close querier for {%s="%s"}`, defaultTenantLabel, tenantID))
 	}
 	return errs.Err()
 }
 
+type selectJob struct {
+	seriesSet *storage.SeriesSet
+	querier   storage.Querier
+	tenantID  string
+}
+
 // Select returns a set of series that matches the given label matchers. If the
 // tenantLabelName is matched on it only considers those queriers matching. The
 // forwarded labelSelector is not containing those that operate on
 // tenantLabelName.
 func (m *mergeQuerier) Select(sortSeries bool, hints *storage.SelectHints, matchers ...*labels.Matcher) storage.SeriesSet {
 	matchedTenants, filteredMatchers := filterValuesByMatchers(defaultTenantLabel, m.tenantIDs, matchers...)
-	var seriesSets = make([]storage.SeriesSet, 0, len(matchedTenants))
-	for pos, tenantID := range m.tenantIDs {
-		if _, matched := matchedTenants[tenantID]; !matched {
+	var jobs = make([]interface{}, len(matchedTenants))
+	var seriesSets = make([]storage.SeriesSet, len(matchedTenants))
+	var jobPos int
+	for tenantPos := range m.tenantIDs {
+		if _, matched := matchedTenants[m.tenantIDs[tenantPos]]; !matched {
 			continue
 		}
-		seriesSets = append(seriesSets, &addLabelsSeriesSet{
-			// TODO: Consider running Select calls concurrently
-			upstream: m.queriers[pos].Select(sortSeries, hints, filteredMatchers...),
+		jobs[jobPos] = &selectJob{
+			seriesSet: &seriesSets[jobPos],
+			querier:   m.queriers[tenantPos],
+			tenantID:  m.tenantIDs[tenantPos],
+		}
+		jobPos++
+	}
+
+	run := func(ctx context.Context, jobIntf interface{}) error {
+		// end early when context has ended
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+
+		job, ok := jobIntf.(*selectJob)
+		if !ok {
+			return fmt.Errorf("unexpected type %T", jobIntf)
+		}
+		*job.seriesSet = &addLabelsSeriesSet{
+			upstream: job.querier.Select(sortSeries, hints, filteredMatchers...),
 			labels: labels.Labels{
 				{
 					Name:  defaultTenantLabel,
-					Value: tenantID,
+					Value: job.tenantID,
 				},
 			},
-		})
+		}
+		return nil
+	}
+
+	err := concurrency.ForEach(m.ctx, jobs, maxConcurrency, run)
+	if err != nil {
+		panic(fmt.Sprintf("this should never happen: %v", err))
```

Comment:
this will happen if context is canceled. We can return `storage.ErrSeriesSet` in that case, with given error.

Reply:
Thanks for pointing `storage.ErrSeriesSet` out, moved using it in a5970a2